src/core/filesystem.cc

```diff
@@ -681,6 +690,351 @@ GCSFileSystem::WriteTextFile(
 
 #endif  // TRITON_ENABLE_GCS
 
+
+#ifdef TRITON_ENABLE_AZURE
+
+namespace as = azure::storage_lite;
+
+class ASFileSystem :  public FileSystem {
+ public:
+  ASFileSystem(const std::string& s3_path);
+  Status CheckClient();
+
+  Status FileExists(const std::string& path, bool* exists);
+  Status IsDirectory(const std::string& path, bool* is_dir);
+  Status FileModificationTime(const std::string& path, int64_t* mtime_ns);
+  Status GetDirectoryContents(
+      const std::string& path, std::set<std::string>* contents);
+  Status GetDirectorySubdirs(
+      const std::string& path, std::set<std::string>* subdirs);
+  Status GetDirectoryFiles(
+      const std::string& path, std::set<std::string>* files);
+  Status ReadTextFile(const std::string& path, std::string* contents);
+  Status LocalizeDirectory(
+      const std::string& path, std::shared_ptr<LocalizedDirectory>* localized);
+  Status WriteTextFile(const std::string& path, const std::string& contents);
+
+ private:
+  Status ParsePath(
+      const std::string& path, std::string* bucket, std::string* object);
+  std::shared_ptr<as::blob_client> client_;
+
+  Status ListDirectory(
+      const std::string& path, const std::string& dir_path,
+      std::function<
+          Status(const as::list_blobs_segmented_item&, const std::string&)>
+          func);
+
+  Status DownloadFolder(
+      const std::string& container, const std::string& path,
+      const std::string& dest);
+  re2::RE2 as_regex_;
+};
+
+Status
+ASFileSystem::ParsePath(
+    const std::string& path, std::string* container, std::string* object)
+{
+  std::string host_name, query;
+  if (!RE2::FullMatch(path, as_regex_, &host_name, container, object, &query)) {
+    return Status(
+        Status::Code::INTERNAL, "Invalid azure storage path: " + path);
+  }
+  return Status::Success;
+}
+
+ASFileSystem::ASFileSystem(const std::string& path)
+    : as_regex_("as://([^/]+)/([^/]+)/([^?]+)(\\?.*)?")
+{
+  const char* account_key = std::getenv("AZURE_STORAGE_KEY");
+  std::shared_ptr<as::storage_account> account = nullptr;
+  std::string host_name, container, blob_path, query;
+  if (RE2::FullMatch(
+          path, as_regex_, &host_name, &container, &blob_path, &query)) {
+    size_t pos = host_name.rfind(".blob.core.windows.net");
+    std::string account_name;
+    if (pos != std::string::npos) {
+      account_name = host_name.substr(0, pos);
+    } else {
+      account_name = host_name;
+    }
```

Comment:
Does this convention have a doc?

Reply:
No. `{account}.blob.core.windows.net` is the default assigned DNS name. But user can assign any domain to its storage access.

Adding a new env `AZURE_STORAGE_ACCOUNT` to cover this situation.